src/app/component/key/key-interface.js

```diff
@@ -0,0 +1,294 @@
+  }
+
+  constructor(keyMapItem) {
+    if (!(keyMapItem instanceof Object)) { throw new Error('wrong keymap item'); }
+    this.keyType = (keyMapItem.keyType === 'spec') ? 'spec' : keyMapItem.keyVal[0].type;
+    this.keyId = keyMapItem.keyCode;
+    this.keyValues = keyMapItem.keyVal;
+    this.keyActive = false;
+    this.char = '';
+    this.options = keyMapItem.options;
+    this.shellElement = createElement({
+      tagName: 'div',
+      className: [CssClasses.BLOCK, CssClasses.SHELL, CssClasses.SH[this.keyType]].join(' '),
+      events: [{ event: 'mousedown', handler: this.click.bind(this) }],
+    });
+
+    /* Basic markup */
+    if (this.keyType === 'spec') {
+      this.descName = createElement({ tagName: 'span', className: [CssClasses.DESCRIPT, CssClasses.DESC.spec].join(' ') });
+
+      this.shellElement.append(this.descName);
+    } else {
+      this.descMain = createElement({ tagName: 'span', className: CssClasses.DESCRIPT });
+      this.descAlt = createElement({ tagName: 'span', className: CssClasses.DESCRIPT });
+
+      this.shellElement.append(this.descMain, this.descAlt);
+    }
+
+    /* apply styles and charset */
+    this.setShellStyle(true);
+    this.setChar();
+
+    KeyInterface.array.push(this);
+  }
+
+  setShellStyle(forceType) {
+    if (forceType || (this.keyType !== 'spec' && this.keyType !== this.keyValues[KeyInterface.lang].type)) {
+      if (this.keyType === 'spec') {
+        this.shellElement.className = [CssClasses.BLOCK, CssClasses.SHELL, CssClasses.SH.spec].join(' ');
+        this.descName.className = [CssClasses.DESCRIPT, CssClasses.DESC.spec].join(' ');
+      } else {
+        this.keyType = this.keyValues[KeyInterface.lang].type;
+
+        this.shellElement.className = [CssClasses.BLOCK, CssClasses.SHELL, CssClasses.SH[this.keyType]].join(' ');
+        this.descMain.className = [CssClasses.DESCRIPT, CssClasses.DESC[`${this.keyType}Main`]].join(' ');
+        this.descAlt.className = [CssClasses.DESCRIPT, CssClasses.DESC[`${this.keyType}Alt`]].join(' ');
+      }
+    }
+
+    if (this.options) {
+      if (this.options.includes('stretch')) {
+        this.shellElement.classList.add(CssClasses.SHELL_STRETCH);
+      }
+      if (this.options.includes('semiwide')) {
+        this.shellElement.classList.add(CssClasses.WIDE);
+      }
+      // media
+      if (this.options.includes('span2')) {
+        this.shellElement.classList.add(CssClasses.SHELL_M_SP2);
+      }
+      if (this.options.includes('span-auto')) {
+        this.shellElement.classList.add(CssClasses.SHELL_M_SPAUTO);
+      }
+    }
+  }
+
+  setChar(shiftCheck) {
+    if (this.keyType === 'spec') {
+      this.descName.textContent = this.keyValues;
+    } else {
+      let swap = false;
+      if (shiftCheck) {
+        if (this.keyType === 'letter') {
+          swap = KeyInterface.flags.caps === !KeyInterface.flags.shift;
+        } else {
+          swap = KeyInterface.flags.shift;
+        }
+      }
```

Comment:
For every place, where you want to write `let` you can create separate method for this logic.
`const swap = this.getSwap(shiftCheck, this.keyType)`